test_t20 of pbs_reliable_job_startup.py is failing while verifying the job's substate

Mahalty opened the pull request:
#### Bug/feature Description
* "test_t20" of TestPbsReliableJobStartup is failing while verifying the job's substate. Expected job's substate is 41 but observed substate of the job is 42

#### Affected Platform(s)
* ALL

#### Cause / Analysis / Design
* In the test before checking for the job's substate we are verifying certain log-msgs in accounting logs which is introducing some delay(Due to sudoer's issue). By the time the test is checking for the job to be in 41 state it already entered substate 42

#### Solution Description
* Check accounting logs after we check the job attributes

#### Testing logs/output
* [Execution_logs_TestPbsReliableJobStartup_test_t20_before_fix.txt](https://github.com/PBSPro/pbspro/files/2863725/Execution_logs_TestPbsReliableJobStartup_test_t20_before_fix.txt)
* [Execution_logs_TestPbsReliableJobStartup_test_t20_after_fix.txt](https://github.com/PBSPro/pbspro/files/2863719/Execution_logs_TestPbsReliableJobStartup_test_t20_after_fix.txt)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
 - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bayucan commented on test/tests/functional/pbs_reliable_job_startup.py:
```diff
@@ -5452,6 +5452,8 @@ def test_t20(self):
             self.momA.log_match("Job;%s;pruned to exec_vnode=%s" % (
                 sjid, exec_vnode), n=10)
+
+        for idx in range(1, 4):
```
Just attaching the outer for loop here would just result in checking accounting_logs for sjid_esc, iexec_host_esc, and iexec_vnode_esc values of the subjob at index 3 from previous for loop, and running that 3 times.
The sjid_esc, iexec_host_esc, and iexec_vnode_esc values from previous loop must be saved as arrays indexed by the subjob index. Then the individual subjob entries can be checked as sjid_esc[idx], iexec_host_esc[idx], and iexec_vnode_esc[idx]...

Mahalty commented:
@anamikau @bayucan  Thanks for the review.
@bayucan I have addressed your review comments. Please review
**Latest execution logs:**
[Execution_logs_TestPbsReliableJobStartup_test_t20_after_fix_1.txt](https://github.com/PBSPro/pbspro/files/2869413/Execution_logs_TestPbsReliableJobStartup_test_t20_after_fix_1.txt)

bayucan commented on test/tests/functional/pbs_reliable_job_startup.py:
```diff
@@ -5374,53 +5374,56 @@ def test_t20(self):
                                id=sjid, attrop=PTL_AND)
 
+        for idx in range(1, 4):
+            sjid = create_subjob_id(jid, idx)
```
Rather than calling create_subjob_id() again here, you can just cache the sjid values in an array, like adding something after line 5411:
 sjid_values[idx] = sjid, 
then can just put in here:
 sjid = sjid_values[idx].

bayucan commented on test/tests/functional/pbs_reliable_job_startup.py:
```diff
@@ -5374,53 +5374,56 @@ def test_t20(self):
-                vnode_list = [self.nAv0, self.nB, self.nC,
-                              self.nD, self.nE]
+                d = {'iexec_host_esc' : self.jobA_iexec_host1_esc,
+                'iexec_vnode' : self.jobA_iexec_vnode1,
```
I think Travis is complaining about indentation here.

Mahalty commented:
@bayucan I have addressed your review comments. Please review

**Latest execution-logs:**
[Execution_logs_TestPbsReliableJobStartup_test_t20_after_fix_2.txt](https://github.com/PBSPro/pbspro/files/2874673/Execution_logs_TestPbsReliableJobStartup_test_t20_after_fix_2.txt)

agrawalravi90 commented:
you mentioned that the issue was that we were doing log_match before checking for substate 41, but it seems like we were checking for substate 41 before we did log_match even before your changes right?

Mahalty commented:
@agrawalravi90 In the test we are checking for job attributes, verifing certain msgs in mom and accounting logs for every subjob (using for loop).  By the time the test is checking all the mentioned things for second subjob (in second iteration) the third subjob started running(i.e substate 42) and the test is expecting job to be in 41 substate,Hence the test failed. To fix this issue, I am creating 2 loops. One for checking the job attributes for all the sub jobs and another loop for checking the logs msgs in accounting and mom logs

Mahalty commented:
@PBSPro/pbspro-maintainers I have squashed my commits and rebased my branch. Please take this PR forward
